internal/gps/pkgtree/digest.go

```diff
@@ -0,0 +1,442 @@
+// non-empty file, the file's contents are incuded. If a non-empty directory,
+// the contents of the directory are included.
+//
+// While filepath.Walk could have been used, that standard library function
+// skips symbolic links, and for now, we want the hash to include the symbolic
+// link referents.
+func DigestFromPathname(prefix, pathname string) ([]byte, error) {
+	// Create a single hash instance for the entire operation, rather than a new
+	// hash for each node we encounter.
+	h := sha256.New()
+
+	// Initialize a work queue with the os-agnostic cleaned up pathname. Note
+	// that we use `filepath.Clean` rather than `filepath.Abs`, because the hash
+	// has pathnames which are relative to prefix, and there is no reason to
+	// convert to absolute pathname for every invocation of this function.
+	prefix = filepath.Clean(prefix) + pathSeparator
+	prefixLength := len(prefix) // store length to trim off pathnames later
+	pathnameQueue := []string{filepath.Join(prefix, pathname)}
+
+	var written int64
+
+	// As we enumerate over the queue and encounter a directory, its children
+	// will be added to the work queue.
+	for len(pathnameQueue) > 0 {
+		// Unshift a pathname from the queue (breadth-first traversal of
+		// hierarchy)
+		pathname, pathnameQueue = pathnameQueue[0], pathnameQueue[1:]
+
+		fi, err := os.Lstat(pathname)
+		if err != nil {
+			return nil, errors.Wrap(err, "cannot Lstat")
+		}
+		mode := fi.Mode()
+
+		// Skip file system nodes we are not concerned with
+		if mode&skipSpecialNodes != 0 {
+			continue
+		}
+
+		// Write the prefix-stripped pathname to hash because the hash is as
+		// much a function of the relative names of the files and directories as
+		// it is their contents. Added benefit is that even empty directories
+		// and symbolic links will effect final hash value. Use
+		// `filepath.ToSlash` to ensure relative pathname is os-agnostic.
+		writeBytesWithNull(h, []byte(filepath.ToSlash(pathname[prefixLength:])))
+
+		if mode&os.ModeSymlink != 0 {
+			referent, err := os.Readlink(pathname)
+			if err != nil {
+				return nil, errors.Wrap(err, "cannot Readlink")
+			}
+			// Write the os-agnostic referent to the hash and proceed to the
+			// next pathname in the queue.
+			writeBytesWithNull(h, []byte(filepath.ToSlash(referent)))
+			continue
+		}
+
+		// For both directories and regular files, we must create a file system
+		// handle in order to read their contents.
+		fh, err := os.Open(pathname)
+		if err != nil {
+			return nil, errors.Wrap(err, "cannot Open")
+		}
+
+		if fi.IsDir() {
+			childrenNames, err := sortedListOfDirectoryChildrenFromFileHandle(fh)
+			if err != nil {
+				_ = fh.Close() // already have an error reading directory; ignore Close result.
+				return nil, errors.Wrap(err, "cannot get list of directory children")
+			}
+			for _, childName := range childrenNames {
+				switch childName {
+				case ".", "..", "vendor", ".bzr", ".git", ".hg", ".svn":
```

Comment:
> i don't think we should skip it here, though.

I agree with this. If `Godeps` were to change and nothing else, it would be a representation of the authors of that package to use a new version of one of its dependencies. Perhaps a new version fixes a bug that manifests in the package we're evaluating. And the fact that the user's package has a dependency, which has its own dependency change represents, at least in my mind, a new desired build state.